autorest/adal/token.go

```diff
@@ -653,6 +658,78 @@ func (spt *ServicePrincipalToken) refreshInternal(resource string) error {
 	return spt.InvokeRefreshCallbacks(token)
 }
 
+func retry(sender Sender, req *http.Request) (resp *http.Response, err error) {
+	retries := []int{
+		http.StatusRequestTimeout,      // 408
+		http.StatusTooManyRequests,     // 429
+		http.StatusInternalServerError, // 500
+		http.StatusBadGateway,          // 502
+		http.StatusServiceUnavailable,  // 503
+		http.StatusGatewayTimeout,      // 504
+	}
+	// Extra retry status codes requered
+	retries = append(retries, http.StatusNotFound,
+		// all remaining 5xx
+		http.StatusNotImplemented,
+		http.StatusHTTPVersionNotSupported,
+		http.StatusVariantAlsoNegotiates,
+		http.StatusInsufficientStorage,
+		http.StatusLoopDetected,
+		http.StatusNotExtended,
+		http.StatusNetworkAuthenticationRequired)
+
+	attempt := 0
+	maxAttempts := 5
+
+	for attempt < maxAttempts {
+		resp, err = sender.Do(req)
+		if err != nil {
+			return
+		}
+
+		if resp.StatusCode == http.StatusOK {
+			return
+		}
+		if containsInt(retries, resp.StatusCode) {
+			delayed := false
+			if resp.StatusCode == http.StatusTooManyRequests {
+				delayed = delay(resp, req.Cancel)
```

Comment:
>delay [](start = 14, length = 5)

we should check the delay for the other status codes also. But lets keep the logic for StatusTooManyRequests to retyr only if the retryAfter header is present.